src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+            }
+
+            return false;
+        }
+
+        internal void Connect(VisualStudioWorkspaceImpl workspace)
+        {
+            this.AssertIsForeground();
+
+            _workspace = workspace;
+
+            var componentModel = workspace.GetVsService<SComponentModel, IComponentModel>();
+            _packageInstallerServices = componentModel.GetService<IVsPackageInstallerServices>();
+            _packageInstaller = componentModel.GetService<IVsPackageInstaller>();
+
+            _workspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        internal void Disconnect(VisualStudioWorkspaceImpl workspace)
+        {
+            this.AssertIsForeground();
+
+            Debug.Assert(workspace == _workspace);
+            _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+            _workspace = null;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            // Can be called on any thread.
+            bool localSolutionChanged = false;
+            ProjectId localChangedProject = null;
+            switch (e.Kind)
+            {
+                default:
+                    // Nothing to do for any other events.
+                    return;
+                case WorkspaceChangeKind.ProjectAdded:
+                case WorkspaceChangeKind.ProjectChanged:
+                case WorkspaceChangeKind.ProjectReloaded:
+                case WorkspaceChangeKind.ProjectRemoved:
+                    localChangedProject = e.ProjectId;
+                    break;
+
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionChanged:
+                case WorkspaceChangeKind.SolutionCleared:
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                    localSolutionChanged = true;
+                    break;
+            }
+
+            lock (_gate)
+            {
+                // Cancel any existing update task.
+                _tokenSource.Cancel();
+                _tokenSource = new CancellationTokenSource();
+
+                _solutionChanged |= localSolutionChanged;
+                if (localChangedProject != null)
+                {
+                    _changedProjects.Add(localChangedProject);
+                }
+
+                // Because we may get a lot of events, we attempt to throttle things so that
+                // we only update our nuget information once a second has passed.
+                var cancellationToken = _tokenSource.Token;
+                Task.Delay(TimeSpan.FromSeconds(1), cancellationToken)
+                    .ContinueWith(_ => OnWorkspaceChangedOnForeground(cancellationToken), cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, this.ForegroundTaskScheduler);
+            }
+        }
+
+        private void OnWorkspaceChangedOnForeground(CancellationToken cancellationToken)
```

Comment:
This isn't simply `OnWorkspaceChanged` on the foreground, it's more like `ProcessBatchedChanges` or something